### PR TITLE
[AMDGPU] Exclude certain opcodes from being marked as single use

### PR DESCRIPTION
**Pull request by ScottEgerton**
The s_singleuse_vdst instruction is used to mark regions of instructions that
produce values that have only one use.
Certain instructions take more than one cycle to execute, resulting in regions
being incorrectly marked.
This patch excludes these multi-cycle instructions from being marked as either
producing single use values or consuming single use values or both depending
on the instruction.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp [hunk 1]
```diff
@@ -132,6 +132,153 @@ class AMDGPUInsertSingleUseVDST : public MachineFunctionPass {
 
   AMDGPUInsertSingleUseVDST() : MachineFunctionPass(ID) {}
 
+  static bool IsValidOpcode(const MachineInstr &MI) {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:135]: This feels like it needs to be an instruction bit and/or searchable table

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp [hunk 2]
```diff
@@ -132,6 +132,153 @@ class AMDGPUInsertSingleUseVDST : public MachineFunctionPass {
...
+          return true;
+        default:
+          return false;
+        }
+      }
+      return true;
+    }
+  }
+
+  static bool IsValidConsumerOpcode(const MachineInstr &MI) {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:176]: Start with lowercase

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:135]: Or could we derive this from the scheduling info?

Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:135]: Searchable table sounds good. We can't derive it from the sched model because it's not really related to how many cycles the instruction takes to execute; we should treat it as a separate list of instructions defined by the hardware design.

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp (hunk 2 above)
Comment (ScottEgerton) [reply llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp]: Done

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp (hunk 1 above)
Comment (ScottEgerton) [reply llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp]: I've changed this to be a searchable table.

File: llvm/lib/Target/AMDGPU/VOPInstructions.td [hunk 3]
```diff
@@ -819,9 +825,7 @@ class VOP3P_DPPe_Common<bits<7> op, VOPProfile P> : VOP3P_DPPe_Common_Base<op, P
 
 class VOP_DPP_Pseudo <string OpName, VOPProfile P, list<dag> pattern=[],
   dag Ins = P.InsDPP, string asmOps = P.AsmDPP> :
-  InstSI <P.OutsDPP, Ins, OpName#asmOps, pattern>,
-  VOP <OpName>,
-  SIMCInstr <OpName#"_dpp", SIEncodingFamily.NONE> {
+  VOP_Pseudo<OpName, "_dpp", P, P.OutsDPP, Ins, asmOps, pattern> {
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/VOPInstructions.td:828]: Why this change - is it just a cleanup?

File: llvm/lib/Target/AMDGPU/VOP3Instructions.td [hunk 4]
```diff
@@ -156,10 +156,12 @@ defm V_MAX_F64 : VOP3Inst <"v_max_f64", VOP3_Profile<VOP_F64_F64_F64>, fmaxnum_l
 } // End SchedRW = [WriteDoubleAdd]
 
 let SchedRW = [WriteIntMul] in {
+defm V_MUL_LO_I32 : VOP3Inst <"v_mul_lo_i32", V_MUL_PROF<VOP_I32_I32_I32>>;
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/VOP3Instructions.td:159]: You don't need to handle V_MUL_LO_I32 differently from the other mul opcodes here. It was never generated because its behaviour was always identical to the U32 version, so recent generations have removed support for it, so it does not matter if you set the IsInvalid flag on it.

File: llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp [hunk 5]
```diff
@@ -607,6 +615,16 @@ bool isTrue16Inst(unsigned Opc) {
   return Info ? Info->IsTrue16 : false;
 }
 
+bool isInvalidSingleUseConsumerInst(unsigned Opc) {
+  const SingleUseExceptionInfo *Info = getSingleUseExceptionHelper(Opc);
+  return Info ? Info->IsInvalidSingleUseConsumer : false;
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:620]: Nit: `return Info && Info->IsInvalidSingleUseConsumer`. (I realize you were just copying the existing code.)

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp [hunk 6]
```diff
@@ -132,6 +133,14 @@ class AMDGPUInsertSingleUseVDST : public MachineFunctionPass {
 
   AMDGPUInsertSingleUseVDST() : MachineFunctionPass(ID) {}
 
+  static bool isValidProducerInst(const MachineInstr &MI) {
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:136]: Probably simpler to inline this function into its only call site. Then we would not have a mix of "isValid" and "isInvalid" functions, which I find a little confusing.

File: llvm/test/CodeGen/AMDGPU/insert-singleuse-vdst.mir [hunk 7]
```diff
@@ -1,6 +1,5 @@
 # NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 4
 # RUN: llc -mtriple=amdgcn -mcpu=gfx1150 -mattr=+wavefrontsize32,-wavefrontsize64 -verify-machineinstrs -run-pass=amdgpu-insert-single-use-vdst %s -o - | FileCheck %s
```
Comment (jayfoad) [inline llvm/test/CodeGen/AMDGPU/insert-singleuse-vdst.mir:2]: If you don't care about the wave size in this test, remove the -mattr option.

**Review by jayfoad — COMMENTED** [review]
I think this looks good overall, thanks.

File: llvm/lib/Target/AMDGPU/VOPInstructions.td (hunk 3 above)
Comment (ScottEgerton) [reply llvm/lib/Target/AMDGPU/VOPInstructions.td:828]: This change allows dpp instructions to be included in the searchable table by inheriting the IsInvalidSingleUse* bits from VOP_Pseudo.

Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/VOPInstructions.td:828]: I see. I think this just allows you to use VOP_Pseudo as a common superclass for all VOP instructions, DPP or not, when you define the searchable table. Perhaps you could use an even more common class like InstSI - I'm not sure if that would make the generated table bigger.

@Sisyph @kosarev any thoughts on this part of the patch?

File: llvm/test/CodeGen/AMDGPU/insert-singleuse-vdst.mir (hunk 7 above)
Comment (ScottEgerton) [reply llvm/test/CodeGen/AMDGPU/insert-singleuse-vdst.mir]: Done

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp (hunk 6 above)
Comment (ScottEgerton) [reply llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp]: Done

File: llvm/lib/Target/AMDGPU/VOP3Instructions.td (hunk 4 above)
Comment (ScottEgerton) [reply llvm/lib/Target/AMDGPU/VOP3Instructions.td]: That's good to know. I've changed the code back to handle all the mul opcodes together.

File: llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp (hunk 5 above)
Comment (ScottEgerton) [reply llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp]: Done

File: llvm/lib/Target/AMDGPU/VOPInstructions.td (hunk 3 above)
Comment (kosarev) [reply llvm/lib/Target/AMDGPU/VOPInstructions.td:828]: Looks reasonable to me and a refinement as well. I second Jay's words on whether it makes sense to limit the flags to `VOP_Pseudo`s and then have explicit `SIInstrInfo::isVAL()` checks, though.

**Review by jayfoad — APPROVED** [review]
LGTM, thanks!

File: llvm/lib/Target/AMDGPU/VOPInstructions.td (hunk 3 above)
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/VOPInstructions.td:828]: As a followup I guess we could remove a lot of the definitions below like `let isPseudo = 1` that can now be inherited from `VOP_Pseudo`.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.td [hunk 8]
```diff
@@ -2266,6 +2266,8 @@ class VOPProfile <list<ValueType> _ArgVT, bit _EnableClamp = 0> {
   field bit EnableClamp = _EnableClamp;
   field bit IsTrue16 = 0;
   field bit IsRealTrue16 = 0;
+  field bit IsInvalidSingleUseConsumer = 0;
+  field bit IsInvalidSingleUseProducer = 0;
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.td:2270]: Looking again, there is no reason for these fields to exist in VOPProfile.

File: llvm/lib/Target/AMDGPU/VOPInstructions.td [hunk 9]
```diff
@@ -175,6 +179,8 @@ class VOP3P_Pseudo <string opName, VOPProfile P, list<dag> pattern = []> :
 class VOP_Real<VOP_Pseudo ps> {
   Instruction Opcode = !cast<Instruction>(NAME);
   bit IsSingle = ps.Pfl.IsSingle;
+  bit IsInvalidSingleUseConsumer = ps.Pfl.IsInvalidSingleUseConsumer;
+  bit IsInvalidSingleUseProducer = ps.Pfl.IsInvalidSingleUseProducer;
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/VOPInstructions.td:183]: Copy from the pseudo not the profile.

File: llvm/lib/Target/AMDGPU/VOPInstructions.td [hunk 10]
```diff
@@ -81,6 +83,8 @@ class VOP_Pseudo <string opName, string suffix, VOPProfile P, dag outs, dag ins,
   string Mnemonic = opName;
   Instruction Opcode = !cast<Instruction>(NAME);
   bit IsTrue16 = P.IsTrue16;
+  bit IsInvalidSingleUseConsumer = P.IsInvalidSingleUseConsumer;
+  bit IsInvalidSingleUseProducer = P.IsInvalidSingleUseProducer;
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/VOPInstructions.td:87]: Just set them to 0 (or false) here and let specific instructions override them.

**Review by jayfoad — COMMENTED** [review]
Just noticed some unnecessary stuff in this patch. Could you submit a follow up to clean it up please?
